src/coprocessor/codec/convert.rs

```diff
@@ -243,23 +404,53 @@ fn get_valid_float_prefix<'a>(ctx: &mut EvalContext, s: &'a str) -> Result<&'a s
             valid_len = i + 1;
         }
     }
-    box_try!(ctx.handle_truncate(valid_len < s.len()));
+    ctx.handle_truncate(valid_len < s.len())?;
     if valid_len == 0 {
         Ok("0")
     } else {
         Ok(&s[..valid_len])
     }
 }
 
+fn round_int_str(num_next_dot: char, s: &str) -> Cow<'_, str> {
+    if num_next_dot < '5' {
+        return Cow::Borrowed(s);
+    }
+
+    let mut int_str = String::new();
+    match s.rfind(|c| c != '9' && c != '+' && c != '-') {
+        Some(idx) => {
+            int_str.push_str(&s[..idx]);
+            let next_char = char::from_u32(s.chars().nth(idx).unwrap() as u32 + 1).unwrap();
```

Comment:
From `nth`'s doc
> Note that all preceding elements, as well as the returned element, will be consumed from the iterator. That means that the preceding elements will be discarded, and also that calling nth(0) multiple times on the same iterator will return different elements.

This is str contain valid decimal, so I think use `&[u8]` but not `&str` is better, and we can index the char in `O(1)` time

Reply:
We cannot get a `char` in &[u8] actually.